File: ui/src/app/business/deploy-plan/zone/zone.ts
import {BaseModel, BaseRequest} from '../../../shared/class/BaseModel';
import {Region} from '../region/region';
import {IpPool} from '../ip-pool/ip-pool';

export class Zone extends BaseModel {
    id: string;
    name: string;
    vars: string;
    credentialId: string;
    cloudVars: {} = {};
    regionName: string;
    provider: string;
    status: string;
    ipPool: IpPool = new IpPool();
}

export class ZoneCreateRequest extends BaseRequest {
    vars: string;
    regionName: string;
    regionID: string;
    cloudVars: {} = {};
    provider: string;
    credentialId: string;
    ipPoolName: string;
}

export class ZoneUpdateRequest extends BaseRequest {
    vars: string;
    regionID: string;
    cloudVars: {} = {};
    ipPoolName: string;
}

export class CloudZoneRequest extends BaseRequest {
    cloudVars: {} = {};
    datacenter: string;
}

export class CloudZone {
    cluster: string;
    networks: [] = [];
    resourcePools: [] = [];
    datastores: [] = [];
    storages: Storage[] = [];
    securityGroups: [] = [];
    networkList: Network[] = [];
    floatingNetworkList: Network[] = [];
    ipTypes: [] = [];
    imageList: Image[] = [];
    switchs: Switch[] = [];
    templates: string[] = [];
}

export class Switch {
    name: string;
    portgroups: string[];
}


export class CloudTemplate {
    imageName: string;
    guestId: string;
}

export class Storage {
    id: string;
    name: string;
}

export class Network {
    id: string;
    name: string;
    subnetList: Subnet[] = [];
}

export class Subnet {
    id: string;
    name: string;
}

export class Image {
    id: string;
    name: string;
}

export class CloudDatastore {
    name: string;
    capacity: number;
    freeSpace: number;
}

File: pkg/dto/zone.go
package dto

import "github.com/KubeOperator/KubeOperator/pkg/model"

type Zone struct {
	model.Zone
	CloudVars  interface{} `json:"cloudVars"`
	RegionName string      `json:"regionName"`
	Provider   string      `json:"provider"`
	IpPoolName string      `json:"ipPoolName"`
	IpPool     IpPool      `json:"ipPool"`
}

type ZoneCreate struct {
	Name         string      `json:"name" validate:"koname,required"`
	CloudVars    interface{} `json:"cloudVars" validate:"required"`
	RegionID     string      `json:"regionID" validate:"required"`
	RegionName   string      `json:"regionName" validate:"required"`
	CredentialId string      `json:"credentialId" validate:"required"`
	IpPoolName   string      `json:"ipPoolName" validate:"required"`
}

type ZoneOp struct {
	Operation string `json:"operation" validate:"required"`
	Items     []Zone `json:"items" validate:"required"`
}

type CloudZoneResponse struct {
	Result interface{} `json:"result"`
}

type CloudZoneRequest struct {
	CloudVars  interface{} `json:"cloudVars" validate:"required"`
	Datacenter string      `json:"datacenter"`
	RegionName string      `json:"RegionName"`
}

type ZoneUpdate struct {
	ID         string      `json:"id" validate:"required"`
	Name       string      `json:"name" validate:"required"`
	CloudVars  interface{} `json:"cloudVars" validate:"required"`
	RegionID   string      `json:"regionID" validate:"required"`
	IpPoolName string      `json:"ipPoolName" validate:"required"`
}

type CloudDatastore struct {
	Name      string `json:"name"`
	Capacity  int    `json:"capacity"`
	FreeSpace int    `json:"freeSpace"`
}
